spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -254,6 +317,32 @@ trait TableFeatureSupport { this: Protocol =>
     }
   }
 
+  /**
+   * Remove writer feature from protocol. To remove a writer feature we only need to
+   * remove it from the writerFeatures set.
+   */
+  private[delta] def removeWriterFeature(targetWriterFeature: WriterFeature): Protocol = {
+    require(targetWriterFeature.isRemovable)
+    copy(writerFeatures = writerFeatures.map(_ - targetWriterFeature.name))
+  }
+
+  /**
+   * Remove feature wrapper for removing either Reader/Writer or Writer features. We assume
+   * the feature exists in the protocol. There is a relevant validation at
+   * [[AlterTableRemoveFeatureDeltaCommand]].
+   */
+  def removeFeature(targetFeature: TableFeature): Protocol = targetFeature match {
+    case f: ReaderWriterFeature =>
+      throw new UnsupportedOperationException(
+        "ReaderWriter feature removal is currently not supported")
+    case f: WriterFeature =>
+      removeWriterFeature(f)
+    // Legacy features not currently supported.
```

Comment:
This isn't quite right. Since `LegacyWriterFeature` extends `WriterFeature`, it would have already matched the case, which would have blown up with `IllegalArgumentException` because it's not removable. Which is bad because it's indirect but also we shouldn't be exposing users to arbitrary exceptions in response to a SQL command.
```suggestion
    case f: WriterFeature with RemovableFeature =>
      removeWriterFeature(f)
```
With this change, `removeWriterFeature` can be made strongly typed (no longer needs a `require` internally), and you can delete the `isRemovable` method (this was the only call site).

Reply:
> This isn't quite right. Since LegacyWriterFeature extends WriterFeature, it would have already matched the case, which would have blown up with IllegalArgumentException because it's not removable. Which is bad because it's indirect but also we shouldn't be exposing users to arbitrary exceptions in response to a SQL command.

This is not correct. LegacyWriterFeature does not extend WriterFeature. This is imo counter-intuitive and it causes issues.

Even if that was the case, non-removable features (legacy or not) are already guarded in `AlterTableDropFeatureCommand`. Furthermore, a legacy feature cannot be removable. This is guarded in the removable trait constructor.

> With this change, removeWriterFeature can be made strongly typed (no longer needs a require internally), and you can delete the isRemovable method that no longer has any callers.

We need to check whether a feature is removable early in `AlterTableDropFeatureDeltaCommand` to stop execution early. This method here is called at the very end of the process. I added a clarification in the description that the function assumes the feature is removable. I also added a requirement to make it clear.